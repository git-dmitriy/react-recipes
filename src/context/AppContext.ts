import { createContext } from 'react';
import { ContextTypes } from '@/appTypes';

export const AppContext = createContext({} as ContextTypes);
