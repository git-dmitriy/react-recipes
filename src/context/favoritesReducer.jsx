import { ADD_TO_FAVORITE, REMOVE_FROM_FAVORITE } from './types';

const handlers = {
  [ADD_TO_FAVORITE]: (state, { payload }) => {
    return {
      favorites: [...state.favorites, payload],
    };
  },
  [REMOVE_FROM_FAVORITE]: (state, { payload }) => ({
    favorites: state.favorites.filter((item) => item.idMeal !== payload.idMeal),
  }),
  DEFAULT: (state) => state,
};

export const favoriteReducer = (state, action) => {
  const handler = handlers[action.type] || handlers.DEFAULT;
  return handler(state, action);
};
