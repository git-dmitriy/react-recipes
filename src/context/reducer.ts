import {
    ADD_TO_FAVORITE,
    REMOVE_FROM_FAVORITE,
    SWITCH_THEME,
    SET_IS_LOADING,
} from '@context/contextTypes.ts';
import {StateTypes, ActionTypes} from '@/appTypes';

export const reducer = (state: StateTypes, action: ActionTypes) => {
    let favorites = null;
    switch (action.type) {
        case ADD_TO_FAVORITE:
            favorites = [...state.favorites, action.payload];
            window.localStorage.setItem('favorites', JSON.stringify(favorites));
            return {...state, favorites};

        case REMOVE_FROM_FAVORITE:
            favorites = state.favorites.filter(
                (item) => item.idMeal !== action.payload.idMeal
            );
            window.localStorage.setItem('favorites', JSON.stringify(favorites));
            return {...state, favorites};

        case SWITCH_THEME:
            window.localStorage.setItem('theme', action.payload);
            return {
                ...state,
                theme: action.payload,
            };

        case SET_IS_LOADING:
            return {
                ...state,
                isLoading: action.payload,
            };

        default:
            return state;
    }
};
