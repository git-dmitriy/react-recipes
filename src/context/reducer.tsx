import { ADD_TO_FAVORITE, REMOVE_FROM_FAVORITE, SWITCH_THEME } from './types';
import { StateTypes, ActionTypes } from 'appTypes';

export const reducer = (state: StateTypes, action: ActionTypes) => {
  let favorites = null;
  switch (action.type) {
    case ADD_TO_FAVORITE:
      favorites = [...state.favorites, action.payload];
      window.localStorage.setItem('favorites', JSON.stringify(favorites));
      return { ...state, favorites };

    case REMOVE_FROM_FAVORITE:
      favorites = state.favorites.filter(
        (item) => item.idMeal !== action.payload.idMeal
      );
      window.localStorage.setItem('favorites', JSON.stringify(favorites));
      return { ...state, favorites };

    case SWITCH_THEME:
      window.localStorage.setItem('theme', action.payload);
      return {
        ...state,
        theme: action.payload,
      };

    default:
      return state;
  }
};
