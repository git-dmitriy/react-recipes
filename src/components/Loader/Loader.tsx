import { AppContext } from 'context/AppContext';
import { useContext } from 'react';

export const Loader = () => {
  const circle = 'h-5 w-5 bg-yellow-400 rounded-full';

  const { state } = useContext(AppContext);
  const { isLoading } = state;

  return (
    <div
      role='alert'
      aria-busy={isLoading ? true : false}
      className={`fixed inset-0 flex justify-center items-center my-20 ${
        isLoading ? 'opacity-100' : 'opacity-0'
      }`}
    >
      <div className='flex'>
        <div className={`${circle} mr-1 animate-bounce`}></div>
        <div className={`${circle} mr-1 animate-bounce200`}></div>
        <div className={`${circle} animate-bounce400`}></div>
      </div>
    </div>
  );
};
