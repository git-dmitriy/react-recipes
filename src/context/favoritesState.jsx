import React, { useReducer } from 'react';
import { ADD_TO_FAVORITE, REMOVE_FROM_FAVORITE } from 'context/types';
import { FavoriteContext } from 'context/favoritesContext';
import { favoriteReducer } from 'context/favoritesReducer';

export const FavoritesState = ({ children }) => {
  const favorites = window.localStorage.getItem('favorites');
  const initialState = {
    favorites: JSON.parse(favorites) || [],
  };

  const [state, dispatch] = useReducer(favoriteReducer, initialState);

  const addToFavorites = (item) => {
    dispatch({
      type: ADD_TO_FAVORITE,
      payload: item,
    });
  };

  const removeFromFavorites = (item) => {
    dispatch({
      type: REMOVE_FROM_FAVORITE,
      payload: item,
    });
  };

  return (
    <FavoriteContext.Provider
      value={{
        addToFavorites,
        removeFromFavorites,
        state,
      }}
    >
      {children}
    </FavoriteContext.Provider>
  );
};
