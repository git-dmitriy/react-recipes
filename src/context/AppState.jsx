import React, { useReducer } from 'react';
import {
  ADD_TO_FAVORITE,
  SWITCH_THEME,
  REMOVE_FROM_FAVORITE,
} from 'context/types';
import { AppContext } from 'context/AppContext';
import { reducer } from 'context/reducer';

export const AppState = ({ children }) => {
  const favorites = window.localStorage.getItem('favorites');
  const initialState = {
    favorites: JSON.parse(favorites) || [],
    theme: window.localStorage.getItem('theme') || 'light',
  };

  const [state, dispatch] = useReducer(reducer, initialState);

  const addToFavorites = (item) => {
    dispatch({
      type: ADD_TO_FAVORITE,
      payload: item,
    });
  };

  const removeFromFavorites = (item) => {
    dispatch({
      type: REMOVE_FROM_FAVORITE,
      payload: item,
    });
  };

  const switchTheme = (mode) => {
    if (mode === 'light' || mode === 'dark') {
      dispatch({
        type: SWITCH_THEME,
        payload: mode,
      });
    }
  };

  return (
    <AppContext.Provider
      value={{
        addToFavorites,
        removeFromFavorites,
        switchTheme,
        state,
      }}
    >
      {children}
    </AppContext.Provider>
  );
};
