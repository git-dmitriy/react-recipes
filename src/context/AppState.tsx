import React, { useReducer } from 'react';
import {
  ADD_TO_FAVORITE,
  SWITCH_THEME,
  REMOVE_FROM_FAVORITE,
  SET_IS_LOADING,
} from 'context/types';
import { AppContext } from 'context/AppContext';
import { reducer } from 'context/reducer';
import { StateTypes, MealItemTypes, ContextTypes } from 'appTypes';

type P = { children: React.ReactNode };

export const AppState = ({ children }: P) => {
  const favorites = window.localStorage.getItem('favorites') || '[]';

  const initialState: StateTypes = {
    favorites: JSON.parse(favorites),
    theme: window.localStorage.getItem('theme') || 'light',
    isLoading: false,
  };

  const [state, dispatch] = useReducer(reducer, initialState);

  const addToFavorites = (item: MealItemTypes) => {
    dispatch({
      type: ADD_TO_FAVORITE,
      payload: item,
    });
  };

  const removeFromFavorites = (item: MealItemTypes) => {
    dispatch({
      type: REMOVE_FROM_FAVORITE,
      payload: item,
    });
  };

  const switchTheme = (mode: string) => {
    if (mode === 'light' || mode === 'dark') {
      dispatch({
        type: SWITCH_THEME,
        payload: mode,
      });
    }
  };

  const setIsLoading = (status = false) => {
    dispatch({
      type: SET_IS_LOADING,
      payload: status,
    });
  };

  const value: ContextTypes = {
    addToFavorites,
    removeFromFavorites,
    switchTheme,
    setIsLoading,
    state,
  };

  return <AppContext.Provider value={value}>{children}</AppContext.Provider>;
};
