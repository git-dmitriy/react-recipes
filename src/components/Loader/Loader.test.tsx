import renderer from 'react-test-renderer';
import { Loader } from './Loader';
import { AppContext } from '@/context/AppContext';
import { it, expect, vi } from 'vitest';


const context = {
    state: {
        favorites: [],
        theme: 'light',
        isLoading: false,
    },
    addToFavorites: vi.fn(),
    removeFromFavorites: vi.fn(),
    switchTheme: vi.fn(),
    setIsLoading: vi.fn(),
};

it('should render transparent loader when loading is false', () => {
    const tree = renderer
        .create(
            <AppContext.Provider value={context}>
                <Loader />
            </AppContext.Provider>
        )
        .toJSON();

    expect(tree).toMatchSnapshot();
});

it('should show loader when loading', () => {
    const context = {
        state: {
            favorites: [],
            theme: 'light',
            isLoading: true,
        },
        addToFavorites: vi.fn(),
        removeFromFavorites: vi.fn(),
        switchTheme: vi.fn(),
        setIsLoading: vi.fn(),
    };

    const tree = renderer
        .create(
            <AppContext.Provider value={context}>
                <Loader />
            </AppContext.Provider>
        )
        .toJSON();

    expect(tree).toMatchSnapshot();
});
