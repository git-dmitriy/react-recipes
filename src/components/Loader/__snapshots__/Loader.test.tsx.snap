// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should render transparent loader when loading is false 1`] = `
<div
  aria-busy={false}
  className="fixed inset-0 flex justify-center items-center my-20 opacity-0"
  role="alert"
>
  <div
    className="flex"
  >
    <div
      className="h-5 w-5 bg-yellow-400 rounded-full mr-1 animate-bounce"
    />
    <div
      className="h-5 w-5 bg-yellow-400 rounded-full mr-1 animate-bounce200"
    />
    <div
      className="h-5 w-5 bg-yellow-400 rounded-full animate-bounce400"
    />
  </div>
</div>
`;

exports[`should show loader when loading 1`] = `
<div
  aria-busy={true}
  className="fixed inset-0 flex justify-center items-center my-20 opacity-100"
  role="alert"
>
  <div
    className="flex"
  >
    <div
      className="h-5 w-5 bg-yellow-400 rounded-full mr-1 animate-bounce"
    />
    <div
      className="h-5 w-5 bg-yellow-400 rounded-full mr-1 animate-bounce200"
    />
    <div
      className="h-5 w-5 bg-yellow-400 rounded-full animate-bounce400"
    />
  </div>
</div>
`;
